pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -142,18 +145,43 @@ func toEnvoyHTTPFilter(wasmPlugin *model.WasmPluginWrapper) *hcm_filter.HttpFilt
 // InsertedExtensionConfigurations returns pre-generated extension configurations added via WasmPlugin.
 func InsertedExtensionConfigurations(
 	wasmPlugins map[extensions.PluginPhase][]*model.WasmPluginWrapper,
-	names []string) []*envoy_config_core_v3.TypedExtensionConfig {
+	names []string, pullSecrets map[string][]byte) []*envoy_config_core_v3.TypedExtensionConfig {
 	result := make([]*envoy_config_core_v3.TypedExtensionConfig, 0)
 	if len(wasmPlugins) == 0 {
 		return result
 	}
 	hasName := sets.NewWith(names...)
 	for _, list := range wasmPlugins {
 		for _, p := range list {
-			if !hasName.Contains(p.ExtensionConfiguration.Name) {
+			if !hasName.Contains(p.ResourceName) {
 				continue
 			}
-			result = append(result, proto.Clone(p.ExtensionConfiguration).(*envoy_config_core_v3.TypedExtensionConfig))
+			wasmExtensionConfig := proto.Clone(p.WasmExtensionConfig).(*extensionsv3.Wasm)
+			// Find the pull secret resource name from wasm vm env variables.
+			// The Wasm extension config should already have a `ISTIO_META_WASM_IMAGE_PULL_SECRET` env variable
+			// at in the VM env variables, with value being the secret resource name. We try to find the actual
+			// secret, and replace the env variable value with it. When ECDS config update reaches the proxy,
+			// agent will extract out the secret from env variable, use it for image pulling, and strip the
+			// env variable from VM config before forwarding it to envoy.
+			envs := wasmExtensionConfig.GetConfig().GetVmConfig().GetEnvironmentVariables().GetKeyValues()
+			secretName := envs[model.WasmSecretEnv]
+			if secretName != "" {
+				if sec, found := pullSecrets[secretName]; found {
+					envs[model.WasmSecretEnv] = string(sec)
+				} else {
+					envs[model.WasmSecretEnv] = ""
```

Comment:
As far as I can tell, currently this is ALWAYS set. Not sure if this is the exact place that causes it though.

Should we change it so that:

* Set and non-empty: successfully got secret
* Set and empty: want secret, but failed. Currently this will fail in the agent with obscure error (401 instead of "failed to fetch secret). It actually probably shouldn't even try to pull the plugin without auth -- a server could return different response to authenticated vs unauthenticated for example. Or we should do what k8s does
* Not set: didn't want a secret at all

Reply:
Yeah makes sense. I've updated the behavior to what you described. Set and empty will not result in fetch. Agent will short circuit and nack in that case.